### PR TITLE
refactor: move filter rewriting to expression construction

### PR DESCRIPTION
**Pull request by cpcloud**
- refactor: rewrite filters at expression construction time
- chore(pandas/dask): implement TableArrayView
- chore: don't rewrite filters with Window operations

### CODE REVIEW

File: ibis/expr/analysis.py [hunk 1]
```diff
@@ -1265,3 +1265,67 @@ def _visit_table(self, expr):
...
+@_rewrite_filter.register(ops.Literal)
+@_rewrite_filter.register(ops.ExistsSubquery)
+@_rewrite_filter.register(ops.NotExistsSubquery)
+@_rewrite_filter.register(ops.Window)
+def _rewrite_filter_subqueries(_, expr, **kwargs):
+    return expr
+
+
+@_rewrite_filter.register(ops.Binary)
+def _rewrite_filter_binary(op, expr, **kwargs):
```
Comment (kszucs) [inline ibis/expr/analysis.py:1297]: Can you add comments describing what each overload does?

File: ibis/expr/types/relations.py [hunk 2]
```diff
@@ -738,7 +738,13 @@ def filter(
         child = self
         for predicate, right in top_ks:
             child = child.semi_join(right, predicate)[child]
-        return an.apply_filter(child, resolved_predicates)
+        return an.apply_filter(
```
Comment (kszucs) [inline ibis/expr/types/relations.py:741]: What exactly happens here? :)

Comment (cpcloud) [reply ibis/expr/types/relations.py:741]: Ultimately a `Selection` is constructed. Is there something specific you're wanting to know about?

File: ibis/expr/analysis.py (hunk 1 above)
Comment (cpcloud) [reply ibis/expr/analysis.py]: Done, added docstrings.

File: ibis/expr/types/relations.py (hunk 2 above)
Comment (cpcloud) [reply ibis/expr/types/relations.py:741]: The point of this rewrite is to turn expressions like `a > a.mean()` into an expression that is meaningful to SQL.

File: ibis/expr/analysis.py [hunk 3]
```diff
@@ -1207,3 +1207,55 @@ def _visit_table(self, expr):
                 and base_table.op() not in self.query_roots
             ):
                 self.foreign_table = expr
+
+
+@functools.singledispatch
```
Comment (kszucs) [inline ibis/expr/analysis.py:1212]: What do you think about creating a new package just for rewrite rules (one file per purpose)?

Each file should have a well defined behaviour on the input expression.

Comment (cpcloud) [reply ibis/expr/analysis.py:1212]: Yeah, I think that makes sense. Let's do it in a follow up.

**Review by kszucs — APPROVED** [review]

**Pull request merged**